File: ai/bot/helper/find_item.py
##
## EPITECH PROJECT, 2025
## B-YEP-400-STG-4-1-zappy-noe.carabin
## File description:
## find_item
##

from ... import ai_print as ap
from ... import ai_threads as at

def find_item_in_look(look_string, item_name, thread_name = "UnknownThread"):
    """Finds an item in look results and returns its tile index. Returns -1 if not found or error."""

    if not look_string or not look_string.startswith('[') or not look_string.endswith(']'):
        ap.safe_print(f"[{thread_name}] CRITICAL PARSE ERROR: Bad Look string format: '{look_string}'")
        at.trigger_exit(f"{thread_name} LookParseFormat")
        return -1

    try:
        tiles = look_string.strip('[]').split(',')
        for i, tile_content in enumerate(tiles):
            if item_name in tile_content.split():
                return i
        return -1
    except Exception as e:
        ap.safe_print(f"[{thread_name}] CRITICAL PARSE ERROR: Failed parsing Look string '{look_string}': {e}")
        at.trigger_exit(f"{thread_name} LookParseException")
        return -1

File: ai/entity/bot/helper/find_item.cpp
/*
** EPITECH PROJECT, 2025
** B-YEP-400-STG-4-1-zappy-noe.carabin
** File description:
** find_item
*/

#include "../../AI.hpp"
#include "../../../utils/utils.hpp"
#include <algorithm>

int ai::entity::AI::determineLookDistance(int idx)
{
    int i = 0;
    int lvl = 0;

    while (i + (2 * lvl + 1) <= idx) {
        i += (2 * lvl + 1);
        ++lvl;
    }
    return lvl;
}

int ai::entity::AI::findItemInLook(const std::string &look, const std::string &item)
{
    ai::utils::debug::Logger &logger = ai::utils::debug::Logger::GetInstance();

    if (look.empty() || look.front() != '[' || look.back() != ']') {
        logger.log("[Error] Bad Look string format: '" + look + "'");
        return -1;
    }

    try {
        std::string inner = look.substr(1, look.size() - 2);
        std::vector<std::string> tiles = ai::utils::string::split(inner, ',');

        for (int i = 0; i < static_cast<int>(tiles.size()); ++i) {
            std::vector<std::string> contents = ai::utils::string::split(tiles[i], ' ');
            if (std::find(contents.begin(), contents.end(), item) != contents.end())
                return i;
        }
        return -2;
    } catch (const std::exception &e) {
        logger.log("[Error] Failed parsing Look string '" + look + "': " + e.what());
        return -1;
    }
}
